### PR TITLE
[ASTImporter] Check visibility of EnumDecl.

### PR DESCRIPTION
**Pull request by balazske**
Similar enum decls in different anonymous namespace were not handled correctly.
Added one unit test (maybe add for non-namespace vs namespace case too?), this discovered that the brace-range was not imported (SourceLocation check failure).

### CODE REVIEW

File: unittests/AST/ASTImporterTest.cpp [hunk 1]
```diff
@@ -4593,6 +4593,25 @@ TEST_P(ASTImporterOptionSpecificTestBase,
   EXPECT_EQ(0u, ToTU->getASTContext().getDiagnostics().getNumWarnings());
 }
 
+TEST_P(ASTImporterOptionSpecificTestBase, EnumsInAnonNamespace) {
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: Is this test related to the brace error?

**Review by martong — COMMENTED** [review]
Would it be possible to separate this into two PR?
1) BraceRange related fix
2) EnumDecl visibility (could this have a test in ASTImporterVisibilityTest.cpp ?)

File: unittests/AST/ASTImporterTest.cpp (hunk 1 above)
Comment (balazske) [reply unittests/AST/ASTImporterTest.cpp]: Brace range is not imported correctly, the new test reveals this problem. (Only import of EnumDecl is changed so the changes are related from this point of view. The BraceRange import is tested only with the new test.)

File: lib/AST/ASTImporter.cpp [hunk 2]
```diff
@@ -2453,10 +2455,11 @@ ExpectedDecl ASTNodeImporter::VisitEnumDecl(EnumDecl *D) {
 
   SourceLocation ToBeginLoc;
   NestedNameSpecifierLoc ToQualifierLoc;
+  SourceRange ToBraceRange;
   QualType ToIntegerType;
-  if (auto Imp = importSeq(
-      D->getBeginLoc(), D->getQualifierLoc(), D->getIntegerType()))
-    std::tie(ToBeginLoc, ToQualifierLoc, ToIntegerType) = *Imp;
+  if (auto Imp = importSeq(D->getBeginLoc(), D->getQualifierLoc(),
```
Comment (martong) [inline lib/AST/ASTImporter.cpp]: Do we still need this hunk of change in this PR? Isn't that handled in #633 ?

Comment (balazske) [reply lib/AST/ASTImporter.cpp]: This part will be removed (by rebase) after #632 is applied.

Comment (martong) [reply lib/AST/ASTImporter.cpp]: I see, but it would make sense to create the PRs independently. I mean it would be nice if it would be possible to apply #633 independently of #632. That would mean we can merge #633 first. Also upstreaming would be more straightforward.

Comment (balazske) [reply lib/AST/ASTImporter.cpp]: #633 can be applied first, then update this PR with force-push to remove unneeded parts.

Comment (martong) [reply lib/AST/ASTImporter.cpp]: OK

**Comment by balazske** [comment]
Updated with new diff and new commit message (same code as before).

**Comment by balazske** [comment]
The check was cancelled (not failed) because it was the same code as already tested.

**Comment by martong** [comment]
> The check was cancelled (not failed) because it was the same code as already tested.

Okay, I thought there was an error.

File: unittests/AST/ASTImporterVisibilityTest.cpp [hunk 3]
```diff
@@ -212,6 +262,11 @@ TEST_P(ImportClassesVisibility, ImportAfter) {
 TEST_P(ImportClassesVisibility, ImportAfterImport) {
   TypedTest_ImportAfterImport();
 }
+// EnumDecl.
+TEST_P(ImportEnumsVisibility, ImportAfter) { TypedTest_ImportAfterWithDef(); }
```
Comment (martong) [inline unittests/AST/ASTImporterVisibilityTest.cpp]: So the EnumDecls are squashed? How is it different than in case of classes? What is the difference between ImportAfter and ImportAfterWithDef? (Probablye ImportAfterWithDef is not the best name, since I had this question.)

Comment (balazske) [reply unittests/AST/ASTImporterVisibilityTest.cpp]: It is not possible to forward declare an EnumDecl, it has only definitions. The "WithDef" means that this function is to be used if the thing to import has a definition (this can be a CXXRecordDecl with definition too). These are merged at import or independently created.

Comment (martong) [reply unittests/AST/ASTImporterVisibilityTest.cpp]: > It is not possible to forward declare an EnumDecl, it has only definitions.

What about C++11 enum classes? AFAIK they can have fwd decls.

File: unittests/AST/ASTImporterVisibilityTest.cpp [hunk 4]
```diff
@@ -49,6 +53,9 @@ auto *AnonV = "namespace { extern int v; }";
 // CXXRecordDecl:
 auto *ExternC = "class X;";
 auto *AnonC = "namespace { class X; }";
+// EnumDecl:
```
Comment (martong) [inline unittests/AST/ASTImporterVisibilityTest.cpp:56]: Could we have a test for C++11 enum classes too?
"enum class E;"

Comment (balazske) [reply unittests/AST/ASTImporterVisibilityTest.cpp:56]: This is another case, I can add it.

Comment (balazske) [reply unittests/AST/ASTImporterVisibilityTest.cpp:56]: The enums are currently always "merged" at import, this is probably not always correct. For example when an enum is declared as friend (but maybe this is not problem here?).

Comment (martong) [reply unittests/AST/ASTImporterVisibilityTest.cpp:56]: Yes, that may not be a problem in case of Enums. Though, on the other hand it would be more consistent to not "merge" enum classes and handle them similarly to RecordDecls.

For now, maybe the best would be to skip the test with the definition and prototype but with a FIXME to indicated that enum decls are merged.

Comment (balazske) [reply unittests/AST/ASTImporterVisibilityTest.cpp:56]: The tests work with the new Xxx"WithDef" kind of tests (will be renamed to "XxxWithMerge"). So the tests can be added, maybe with FIXME (at the test or at the import code?).

Comment (balazske) [reply unittests/AST/ASTImporterVisibilityTest.cpp:56]: Still more fixes are needed: If there is an existing enum class fwd declaration the definition is not imported. Should this go into a new PR? (The various fixes of enum class import can be taken as one thing in one PR.)

Comment (martong) [reply unittests/AST/ASTImporterVisibilityTest.cpp:56]: Okay, so let's pretend in this PR that enum classes are nonexistent. And then create a separate PR which fixes enum classes related errors.

Comment (balazske) [reply unittests/AST/ASTImporterVisibilityTest.cpp:56]: Could you accept it now?
The un-scoped enum is handled here.

**Review by martong — APPROVED** [review]
Thanks! Looks good! And nice work :)

**Comment by balazske** [comment]
#634 is the next, for "enum classes".

**Comment by balazske** [comment]
upstream: https://reviews.llvm.org/D62484
